File: app/src/main/java/com/chad/baserecyclerviewadapterhelper/PullToRefreshUseActivity.java
package com.chad.baserecyclerviewadapterhelper;

import android.graphics.Color;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.chad.baserecyclerviewadapterhelper.adapter.PullToRefreshAdapter;
import com.chad.baserecyclerviewadapterhelper.base.BaseActivity;
import com.chad.baserecyclerviewadapterhelper.data.DataServer;
import com.chad.baserecyclerviewadapterhelper.entity.Status;
import com.chad.baserecyclerviewadapterhelper.loadmore.CustomLoadMoreView;
import com.chad.library.adapter.base.listener.OnLoadMoreListener;

import java.util.List;


interface RequestCallBack {
    void success(List<Status> data);

    void fail(Exception e);
}

class Request extends Thread {
    private static final int             PAGE_SIZE = 6;
    private              int             mPage;
    private              RequestCallBack mCallBack;
    private              Handler         mHandler;

    private static boolean mFirstPageNoMore;
    private static boolean mFirstError = true;

    public Request(int page, RequestCallBack callBack) {
        mPage = page;
        mCallBack = callBack;
        mHandler = new Handler(Looper.getMainLooper());
    }

    @Override
    public void run() {
        try {Thread.sleep(500);} catch (InterruptedException e) {}

        if (mPage == 2 && mFirstError) {
            mFirstError = false;
            mHandler.post(new Runnable() {
                @Override
                public void run() {
                    mCallBack.fail(new RuntimeException("fail"));
                }
            });
        } else {
            int size = PAGE_SIZE;
            if (mPage == 1) {
                if (mFirstPageNoMore) {
                    size = 1;
                }
                mFirstPageNoMore = !mFirstPageNoMore;
                if (!mFirstError) {
                    mFirstError = true;
                }
            } else if (mPage == 4) {
                size = 1;
            }

            final int dataSize = size;
            mHandler.post(new Runnable() {
                @Override
                public void run() {
                    mCallBack.success(DataServer.getSampleData(dataSize));
                }
            });
        }
    }
}

/**
 * https://github.com/CymChad/BaseRecyclerViewAdapterHelper
 */
public class PullToRefreshUseActivity extends BaseActivity {

    private static final int PAGE_SIZE = 6;

    private RecyclerView         mRecyclerView;
    private SwipeRefreshLayout   mSwipeRefreshLayout;
    private PullToRefreshAdapter mAdapter;

    private int mNextRequestPage = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        setTitle("Pull TO Refresh Use");
        setBackBtn();

        mRecyclerView = findViewById(R.id.rv_list);
        mSwipeRefreshLayout = findViewById(R.id.swipeLayout);
        mSwipeRefreshLayout.setColorSchemeColors(Color.rgb(47, 223, 189));
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));

        initAdapter();
        addHeadView();
        initRefreshLayout();
        mSwipeRefreshLayout.setRefreshing(true);
        refresh();
    }

    private void initAdapter() {
        mAdapter = new PullToRefreshAdapter();
        mAdapter.getLoadMoreModule().setOnLoadMoreListener(new OnLoadMoreListener() {
            @Override
            public void onLoadMore() {
                loadMore();
            }
        });
        mAdapter.setAnimationEnable(true);
//        mAdapter.setPreLoadNumber(3);
        mRecyclerView.setAdapter(mAdapter);

    }

    private void addHeadView() {
        View headView = getLayoutInflater().inflate(R.layout.head_view, (ViewGroup) mRecyclerView.getParent(), false);
        headView.findViewById(R.id.iv).setVisibility(View.GONE);
        ((TextView) headView.findViewById(R.id.tv)).setText("change load view");
        headView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mAdapter.setNewData(null);
                mAdapter.getLoadMoreModule().setLoadMoreView(new CustomLoadMoreView());
                mRecyclerView.setAdapter(mAdapter);
                Toast.makeText(PullToRefreshUseActivity.this, "change complete", Toast.LENGTH_LONG).show();

                mSwipeRefreshLayout.setRefreshing(true);
                refresh();
            }
        });
        mAdapter.addHeaderView(headView);
    }

    private void initRefreshLayout() {
        mSwipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                refresh();
            }
        });
    }

    private void refresh() {
        mNextRequestPage = 1;
        mAdapter.getLoadMoreModule().setEnableLoadMore(false);//这里的作用是防止下拉刷新的时候还可以上拉加载
        new Request(mNextRequestPage, new RequestCallBack() {
            @Override
            public void success(List<Status> data) {
                setData(true, data);
                mAdapter.getLoadMoreModule().setEnableLoadMore(true);
                mSwipeRefreshLayout.setRefreshing(false);
            }

            @Override
            public void fail(Exception e) {
                Toast.makeText(PullToRefreshUseActivity.this, R.string.network_err, Toast.LENGTH_LONG).show();
                mAdapter.getLoadMoreModule().setEnableLoadMore(true);
                mSwipeRefreshLayout.setRefreshing(false);
            }
        }).start();
    }

    private void loadMore() {
        new Request(mNextRequestPage, new RequestCallBack() {
            @Override
            public void success(List<Status> data) {
                /**
                 * fix https://github.com/CymChad/BaseRecyclerViewAdapterHelper/issues/2400
                 */
                boolean isRefresh =mNextRequestPage ==1;
                setData(isRefresh, data);
            }

            @Override
            public void fail(Exception e) {
                mAdapter.getLoadMoreModule().loadMoreFail();
                Toast.makeText(PullToRefreshUseActivity.this, R.string.network_err, Toast.LENGTH_LONG).show();
            }
        }).start();
    }

    private void setData(boolean isRefresh, List data) {
        mNextRequestPage++;
        final int size = data == null ? 0 : data.size();
        if (isRefresh) {
            mAdapter.setNewData(data);
        } else {
            if (size > 0) {
                mAdapter.addData(data);
            }
        }
        if (size < PAGE_SIZE) {
            //第一页如果不够一页就不显示没有更多数据布局
            mAdapter.getLoadMoreModule().loadMoreEnd(isRefresh);
            Toast.makeText(this, "no more data", Toast.LENGTH_SHORT).show();
        } else {
            mAdapter.getLoadMoreModule().loadMoreComplete();
        }
    }
}

File: kotlinapp/src/main/java/com/allen/kotlinapp/PullToRefreshUseActivity.kt
package com.allen.kotlinapp

import android.graphics.Color
import android.os.Bundle
import android.os.Handler
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import com.allen.kotlinapp.adapter.PullToRefreshAdapter
import com.allen.kotlinapp.base.BaseActivity
import com.allen.kotlinapp.data.DataServer
import com.allen.kotlinapp.loadmore.CustomLoadMoreView
import com.chad.library.adapter.base.BaseQuickAdapter
import com.chad.library.adapter.base.listener.OnItemClickListener

/**
 * 文 件 名: PullToRefreshUseActivity
 * 创 建 人: Allen
 * 创建日期: 2017/6/20 11:24
 * 修改时间：
 * 修改备注：
 */
class PullToRefreshUseActivity : BaseActivity(), BaseQuickAdapter.RequestLoadMoreListener, SwipeRefreshLayout.OnRefreshListener {
    private var mRecyclerView: RecyclerView? = null
    private var pullToRefreshAdapter: PullToRefreshAdapter? = null
    private var mSwipeRefreshLayout: SwipeRefreshLayout? = null

    private val delayMillis = 1000

    private var mCurrentCounter = 0

    private var isErr: Boolean = false
    private var mLoadMoreEndGone = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)
        mRecyclerView = findViewById(R.id.rv_list) as RecyclerView
        mSwipeRefreshLayout = findViewById(R.id.swipeLayout) as SwipeRefreshLayout
        mSwipeRefreshLayout?.setOnRefreshListener(this)
        mSwipeRefreshLayout?.setColorSchemeColors(Color.rgb(47, 223, 189))
        mRecyclerView?.layoutManager = LinearLayoutManager(this)
        setTitle("Pull TO Refresh Use")
        setBackBtn()
        initAdapter()
        addHeadView()
    }

    private fun addHeadView() {
        val headView = getLayoutInflater().inflate(R.layout.head_view, mRecyclerView?.parent as ViewGroup, false)
        headView.findViewById(R.id.iv).setVisibility(View.GONE)
        (headView.findViewById(R.id.tv) as TextView).text = "change load view"
        headView.setOnClickListener(View.OnClickListener {
            mLoadMoreEndGone = true
            pullToRefreshAdapter?.setLoadMoreView(CustomLoadMoreView())
            mRecyclerView?.adapter = pullToRefreshAdapter
            Toast.makeText(this@PullToRefreshUseActivity, "change complete", Toast.LENGTH_LONG).show()
        })
        pullToRefreshAdapter?.addHeaderView(headView)
    }

    override fun onLoadMoreRequested() {
        mSwipeRefreshLayout?.isEnabled = false
        if (pullToRefreshAdapter?.getData()?.size!! < PAGE_SIZE) {
            pullToRefreshAdapter?.loadMoreEnd(true)
        } else {
            if (mCurrentCounter >= TOTAL_COUNTER) {
                //                    pullToRefreshAdapter.loadMoreEnd();//default visible
                pullToRefreshAdapter?.loadMoreEnd(mLoadMoreEndGone)//true is gone,false is visible
            } else {
                if (isErr) {
                    pullToRefreshAdapter?.addData(DataServer.getSampleData(PAGE_SIZE))
                    mCurrentCounter = pullToRefreshAdapter?.getData()?.size!!
                    pullToRefreshAdapter?.loadMoreComplete()
                } else {
                    isErr = true
                    Toast.makeText(this@PullToRefreshUseActivity, R.string.network_err, Toast.LENGTH_LONG).show()
                    pullToRefreshAdapter?.loadMoreFail()

                }
            }
            mSwipeRefreshLayout?.isEnabled = true
        }
    }

    override fun onRefresh() {
        pullToRefreshAdapter?.setEnableLoadMore(false)
        Handler().postDelayed({
            pullToRefreshAdapter?.setNewData(DataServer.getSampleData(PAGE_SIZE))
            isErr = false
            mCurrentCounter = PAGE_SIZE
            mSwipeRefreshLayout?.isRefreshing = false
            pullToRefreshAdapter?.setEnableLoadMore(true)
        }, delayMillis.toLong())
    }

    private fun initAdapter() {
        pullToRefreshAdapter = PullToRefreshAdapter()
        pullToRefreshAdapter?.setOnLoadMoreListener(this, mRecyclerView)
        pullToRefreshAdapter?.openLoadAnimation(BaseQuickAdapter.SLIDEIN_LEFT)
        //        pullToRefreshAdapter.setPreLoadNumber(3);
        mRecyclerView?.adapter = pullToRefreshAdapter
        mCurrentCounter = pullToRefreshAdapter?.getData()?.size!!

        mRecyclerView?.addOnItemTouchListener(object : OnItemClickListener() {
            override fun onSimpleItemClick(adapter: BaseQuickAdapter<*, *>, view: View, position: Int) {
                Toast.makeText(this@PullToRefreshUseActivity, Integer.toString(position), Toast.LENGTH_LONG).show()
            }
        })
    }

    companion object {

        private val TOTAL_COUNTER = 18

        private val PAGE_SIZE = 6
    }


}
